beacon-chain/blockchain/service.go

```diff
@@ -306,7 +307,11 @@ func (c *ChainService) blockProcessing(done <-chan struct{}) {
 			var processedAttestations []*pb.AttestationRecord
 			for index, attestation := range block.Attestations() {
 				// Don't add invalid attestation to block vote cache.
-				if err := c.chain.processAttestation(index, block); err == nil {
+				if err := c.chain.processAttestation(index, block); err != nil {
+					canProcessAttestations = false
+					log.Errorf("could not process attestation for block %d because %v", block.SlotNumber(), err)
+				} else {
+					canProcessAttestations = true
 					processedAttestations = append(processedAttestations, attestation)
 					blockVoteCache, err = c.chain.calculateBlockVoteCache(index, block)
```

Comment:
Why are we stilll updating `blockVoteCache`? Wouldn't the state become inconsistent if we updated the cache but not the active state?

Reply:
It won't be inconsistent, we skip updating the active state if `canProcessAttestations` is false:
```go
	if !canProcessAttestations {
		continue
	}
```
And we only update `blockVoteCache` only if we can`processAttestation` for this particular attestation

Reply:
Ah, I probably could've worded the question better. I think there's a subtle bug here, which is that `calculateBlockVoteCache` modifies the existing cache instead of returning a new copy. Therefore if a block has 10 valid attestations and 1 invalid attestation, we invalidate the entire block but the cache will still record the 10 valid attestations.

Maybe attestation validation should happen in `CanProcessBlock`, and this section can loop through every attestation with the assumption that they've all been validated.